ops/testing.py

```diff
@@ -1099,12 +1102,15 @@ def __init__(self, unit_name, meta):
         # <ID1>: device id that is key for given storage_name
         # Initialize the _storage_list with values present on metadata.yaml
         self._storage_list = {k: {} for k in self._meta.storages}
+
         self._storage_detached = {k: set() for k in self._meta.storages}
         self._storage_id_counter = 0
         # {socket_path : _TestingPebbleClient}
         # socket_path = '/charm/containers/{container_name}/pebble.socket'
         self._pebble_clients = {}  # type: {str: _TestingPebbleClient}
         self._pebble_clients_can_connect = {}  # type: {_TestingPebbleClient: bool}
+        # {<container_name>: <client>}
+        self._pebble_containers = {}  # type: {str: _TestingPebbleClient}
```

Comment:
How distinct is this from `self._pebble_clients` - these seem very closely named and the type annotations seem to suggest a pretty similar structure, too!

Reply:
Reading a bit further down about how this is used, you could almost certainly derive the right value from the socket path that's used as the key in the existing dict? I don't know if that's actually better, just feels like we're duplicating here.

Reply:
I could easily be convinced either way.  Relevant considerations:

* At least one other place in the code uses the pebble_clients dict by rebuilding a socket path starting from a container name.  This makes it seem like a container_name->client mapping might be more natural.
* our socket path conventions could change (but unlikely)
* non-standard socket paths get used in testing.

The last point could be easily remedied with test mods.  If we believe the first point, then it would be worth "fixing" this spot: https://github.com/canonical/operator/blob/e3bc30bf8491716d33c8bb6061eb63bd8306ca1d/ops/testing.py#L759

Reply:
Okay so point 1 makes me wonder whether there is a drive-by refactor in order here? I'm not hard over either way, this shouldn't block merging.

Reply:
This is just food for thought and a general thing we might consider to start doing everywhere:
I recently heard a rather convincing talk by one of the core python devs who recommended doing the following:

Instead of:
```python
 # {<container_name>: <client>}
 self._pebble_containers = {}  # type: {str: _TestingPebbleClient}
```
we could do:

```python
# somewhere on top of the file:
ContainerName = str
# ...

self._pebble_containers: Dict[ContainerName: _TestingPebbleClient] = {}
```

I quite like it.
About this specific line I +1 what's already been said above.

Reply:
I consolidated the dictionaries to one with a container name index and updated all relevant code.